src/backends/perf_pt/collect.c

```diff
@@ -142,6 +163,84 @@ bool perf_pt_start_tracer(struct tracer_ctx *, struct perf_pt_trace *, struct pe
 bool perf_pt_stop_tracer(struct tracer_ctx *tr_ctx, struct perf_pt_cerror *);
 bool perf_pt_free_tracer(struct tracer_ctx *tr_ctx, struct perf_pt_cerror *);
 
+
+/*
+ * Reads samples from the data buffer and acts upon them appropriately.
+ *
+ * Returns true on success, or false otherwise.
+ */
+static bool
+handle_sample(void *aux_buf, struct perf_event_mmap_page *hdr,
+         struct perf_pt_trace *trace, void *data_tmp,
+         struct perf_pt_cerror *err)
+{
+    bool ret = true;
+
+    // We use atomic loads and stores with orderings to avoid concurrency
+    // issues WRT data buffer state mutably shared between us and the kernel.
+    // Without these, for example the compiler might be able to load data
+    // from `data_buf` before reading from the `hdr->data_head`.
+    void *data = (void *) hdr + hdr->data_offset;
+    __u64 head_monotonic =
+            atomic_load_explicit((_Atomic __u64 *) &hdr->data_head,
+                                 memory_order_acquire);
+    __u64 size = hdr->data_size; // No atomic load. Not mutated in-kernel.
+    __u64 head = head_monotonic % size; // Head must be manually wrapped.
+    __u64 tail = atomic_load_explicit((_Atomic __u64 *) &hdr->data_tail,
+                                 memory_order_acquire);
+
+    // Copy samples out, removing wrap in the process.
+    void *data_tmp_end = data_tmp;
+    if (tail <= head) {
+        // Not wrapped.
+        memcpy(data_tmp, data + tail, head - tail);
+        data_tmp_end += head - tail;
+    } else {
+        // Wrapped.
+        memcpy(data_tmp, data + tail, size - tail);
+        data_tmp_end += size - tail;
+        memcpy(data_tmp + size - tail, data, head);
+        data_tmp_end += head;
+    }
+    atomic_store_explicit((_Atomic __u64 *) &hdr->data_tail, head, memory_order_release);
+
+    void *next_sample = data_tmp;
+    while (next_sample != data_tmp_end) {
+        struct perf_event_header *sample_hdr = next_sample;
+        struct perf_record_aux_sample *rec_aux_sample;
+        switch (sample_hdr->type) {
+        case PERF_RECORD_AUX:
+                // Data was written to the AUX buffer.
+                rec_aux_sample = next_sample;
+                // Check that the data written into the AUX buffer was not
+                // truncated. If it was, then we didn't read out of the data buffer
+                // quickly/frequently enough.
+                if (rec_aux_sample->flags & PERF_AUX_FLAG_TRUNCATED) {
+                    perf_pt_set_err(err, perf_pt_cerror_ipt, pte_overflow);
+                    ret = false;
+                    goto clean;
+                }
+                if (read_aux(aux_buf, hdr, trace, err) == false) {
+                    return false;
```

Comment:
Not even using the `goto` label ;)

Reply:
Oops. Well, we are killing the label, so this is actually OK. But we shouldn't mix for sure.